### PR TITLE
Refactor iter

### PR DESCRIPTION
**Pull request by icristescu**
To apply the changes suggested in https://github.com/mirage/irmin/pull/912

### CODE REVIEW

File: src/irmin/irmin.mli [hunk 1]
```diff
@@ -1740,16 +1740,17 @@ module Private : sig
...
 
-      val iter_on_closure :
+      val iter :
         [> `Read ] t ->
         min:node list ->
         max:node list ->
-        f_nodes:(node -> unit Lwt.t) ->
-        f_edges:(node -> node list -> unit Lwt.t) ->
+        ?node:(node -> unit Lwt.t) ->
+        ?edge:(node -> node list -> unit Lwt.t) ->
```
Comment (samoht) [inline src/irmin/irmin.mli]: why `node -> node list -> unit` and not `node -> node -> unit` ?

Comment (icristescu) [reply src/irmin/irmin.mli]: I changed to `?edge:(node -> pred:node -> unit Lwt.t)` as the edges are directed. 
`pred` is clear enough in the context of `object_graph`, where a predecessor relation is also provided as argument to `iter`. But maybe here `~from` and `~to` are better instead?

Comment (samoht) [reply src/irmin/irmin.mli]: `from` and `to` works, or `~src` and `dst`.

**Comment by samoht** [comment]
The API is looks simple and great now, thanks :-)

**Comment by CraigFe** [comment]
Thanks!

File: src/irmin/object_graph.ml [hunk 2]
```diff
@@ -156,32 +157,37 @@ struct
...
           else if has_mark key then visit ()
-          else (
-            mark key level;
-            Log.debug (fun f -> f "VISIT %a %d" Type.(pp X.t) key level);
-            f_nodes key >>= fun () ->
-            pred key >>= fun keys ->
-            List.iter (fun k -> Queue.push (k, level + 1) todo) keys;
-            f_edges key keys >>= fun () -> visit () )
+          else
+            skip key >>= function
```
Comment (samoht) [inline src/irmin/object_graph.ml]: it's probably a good idea to mark the node here so `skip` is called only once for every node.

**Pull request merged**